ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,524 @@
+          new SnapshotDirectorPartitionStep(), // snapshot director which takes regulary snapshots
+          new RocksDbMetricExporterPartitionStep(),
+          new ExporterSatellitePartitionStep() // in order to receive exporter updates
+          );
+```
+
+This should be seen as an example, how it is actually done in the end is implementation detail. The key point is that we need to install some components once, and need to restore/reset some services on transitioning to a new role. We need to make sure that services which depend on each other have the latest reference. Like the SnapshotDirector, should take a snapshot of the actual database object etc. We need to make sure that all services are closed correctly. The ZeebePartition need to support bootstrap steps, which are installed only once. It has to take them done, if the partition is closed.
+
+### Follower Install Request Handling
+
+In order to keep it simple we handle new `InstallRequests` as a new Follower transition, which will cause to recreate all dependent services/components. Why an `InstallRequest` can happen, you can read about it [here](#re-init-follower-partition). How the recreation look like, is again more of an implementation detail. It can be done by calling the `RoleTransitionListeners` again on the Raft level. This approach was used in our [POC](https://github.com/camunda-cloud/zeebe/issues/7328#issuecomment-867683162) and seems to be the simplest one.
+
+       
+## Compatibility
+
+We see no issues with this proposal.
+
+### Rolling updates
+
+We can have various starting scenarios, like followers have the same log, one follower is lagging behind, all followers are lagging behind. When thinking about rolling updates we need to consider how these different scenarios affect the procedure and outcome. 
+
+#### Same state on all nodes
+
+If we have the same state on all nodes and we start a rolling update we could start on a Leader or a Follower.
+
+If we update first the Leader, a Follower will become next Leader, and the old Leader becomes Follower. The new Leader will replicate old entries, which should the Follower with the newest version be able to read in respect to backwards compatibility.
+
+If we update first a Follower, it is similar to above.
+
+#### Follower is lagging behind
+
+If one follower is lagging behind it is likely to happen that if the other Follower or Leader is updated that one of them becomes Leader again. This means we have an old Follower and maybe a Leader with a newer version. In our processing and replay state machine we have event filters, which verify the version of the read entry. Newer entries are ignored by the Follower with the old version. This means for a short period of time the Follower, with the old version, will not replay. This should be fine, since we expect that the rolling update is in a limited time frame and this follower wouldn't take over anyway (since it is lagging behind). 
```

Comment:
> In our processing and replay state machine we have event filters, which verify the version of the read entry. 

Yes, we have a filter in the processing and the replay state machine. But it's good to know that it checks the version of the protocol (defined in `/protocol/pom.xml`) and not the version of the broker itself. 

Currently, the protocol version is not updated strictly. For example, when adding a new value type to the enum.

Apart from changing the protocol definition itself, we might need to consider other changes in the broker. For example, adding a new BPMN element type. If an old follower is not aware of a new BPMN element then it is not able to produce the correct state.

We should think about checking the broker version instead.

> Newer entries are ignored by the Follower with the old version.

Ignored is maybe not the correct word :sweat_smile: The version check throws an exception. I'm not sure how nice this looks in the logs. 

Maybe, we want to handle this case more clear to identify it in the logs when it happens.

Reply:
Thanks I created https://github.com/camunda-cloud/zeebe/issues/7449 for that